Add: optional 'multiline' parameter to perm*Trigger functions

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Add: optional 'multiline' parameter to perm*Trigger functions
#### Motivation for adding to Mudlet
This enables the user to have control over the differences in behaviour. At the moment, triggers created at the root level automatically get 'multiline' enabled and triggers below root level don't get it enabled, and the player has no choice in this matter.
#### Other info (issues closed, discussion etc)
Fixes https://github.com/Mudlet/Mudlet/issues/7480

vadi2 committed:
Add optional 'multiline' parameter to perm*Trigger functions

This enables the user to have control over the differences in behaviour. At the moment, triggers created at the root level automatically get 'multiline' enabled and triggers below root level don't get it enabled, and the player has no choice in this matter.

SlySven commented on src/TLuaInterpreterMudletObjects.cpp:
```diff
@@ -1128,7 +1128,8 @@ int TLuaInterpreter::permRegexTrigger(lua_State* L)
 
     const QString script{lua_tostring(L, 4)};
-    auto [triggerId, message] = pLuaInterpreter->startPermRegexTrigger(name, parent, regList, script);
+    bool multiline = lua_isboolean(L, 5) ? lua_toboolean(L, 5) : false;
```
Given that the fifth argument is optional - does Lua count a missing argument as a "falsey" value - i.e. will `lua_isboolean(...)` on a nil produce an error or coerce it to `false`. If it doesn't and this is a valid means to detect its value if provided - then don't we have to default to a `true` if there is more than one item in the table of "items" for a "root" level trigger so that we maintain our ***backward-compatiblity*** (behaviour) promise?

Also, don't we have a `getVerifiedBool(...)` method to (error) handle some of this (though it doesn't IIRC handle the presence/absence of an argument so that has to be gated behind a `lua_gettop(L)` test?

vadi2 replied:
`nil` in Lua is false, so no backwards compatibility is broken. This can be tested by trying the build available here :wink: 

Adjusted to use `getVerifiedBool` instead.

SlySven replied:
If the user does not supply that fifth "multi-line" option (as previous scripts would not do) then if they supply more than one "item" then prior behaviour would make it a multi-line trigger rather than a multi-item one. So doesn't the option have to default to a `true` (multi-line AND trigger) rather than a not `false` (multi-item OR trigger) if it is not given?

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -6287,6 +6287,7 @@ std::pair<int, QString> TLuaInterpreter::startPermRegexTrigger(const QString& na
     pT->setIsFolder(patterns.empty());
     pT->setIsActive(true);
     pT->setTemporary(false);
+    pT->setIsMultiline(multiline);
```
The line on 6278 should be revised - and indeed the `TTrigger` constructor used there should be modified completely to remove the fourth argument - which is used to set the multi-line mode but which would only be required in the future for the one (of the eleven) uses of that constructor (in `TLuaInterpreter::tempComplexRegexTrigger(...)`) and which could reasonably be done with a line like this there.

This applies to all the Lua API trigger functions revised herein.

vadi2 replied:
I don't think I follow what you are saying.

SlySven replied:
The fourth argument to the `TTrigger(const QString&, const QStringList&, const QList<int>&, bool, Host*)` constructor IS used to set the `(bool) mIsMultiline` in the constructor list - so there is no need to set it in that separate `pT->setIsMultiLine(bool)` call if you do it in the contructor. OTOH if you retain that separate call then there is no need to have it set by that constructor in any of the ***one*** other case outside of these three modified functions - all the others have it permanently set to `false`.

vadi2 committed:
Use getVerifiedBool instead

SlySven commented:
@vadi2 please look over my suggested revisions to this PR - they address the points I've been trying to make (and one or two others)!

vadi2 commented:
Done!

SlySven committed:
Add multiline param triggers (#7498)

#### Brief overview of PR changes/additions
My "improvements" to #7493.

#### Motivation for adding to Mudlet
My perceptions of things that could be done differently/better.

#### Other info (issues closed, discussion etc)
This is to be merged into another PR - not into the `development branch
directly.

---------

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into add-multiline-param-triggers

SlySven commented:
:information_source: The new parameter is an optional 5th one that is an integer - which for a multi-item ("OR") trigger should be a negative number (`-1` is suggested), for a value of `0` or more this makes the trigger a multi-line ("AND") trigger and the number represents the same one as would be entered into the spin-box in the editor and which is enabled only when the "AND/multi-line" groupbox that contains it is checked.

If the optional argument to the `permXxxxTriggers(...)` is not provided the previous "odd" behaviour that the associated issue is reporting is reproduced so as to not change the way Mudlet behaves so that existing scripts work the same.

SlySven commented:
Let's get @NobodyWantsThis to suck this and see - if it *works for them*...

vadi2 commented:
Sounds like existing behaviour is adversely affected https://github.com/Mudlet/Mudlet/issues/7480#issuecomment-2439636658

vadi2 committed:
Merge branch 'development' into add-multiline-param-triggers

SlySven commented:
I've finally got a chance to look at the behaviour of this PR as of 7e36ba5e57 and I compared it to the development branch as of 1fa61b1d8f. I used the following script to generated a trigger group and to create two triggers in it and a further two independent one at the "root level":
```lua
function makeMultiline()
  myTestFunctions = myTestFunctions or {}
  if exists("MyTriggerGroup", "trigger") < 1 then
    -- Have to make two calls to this function as can't insert things into a new
    -- group in the same call!
    myTestFunctions.group = permRegexTrigger("MyTriggerGroup", "", {}, "")
    return
  end

  myTestFunctions.testTrigger1 = permBeginOfLineStringTrigger("dummy1", "", {"Does this work"}, "echo('Yes it does!\\n')")
  myTestFunctions.testTrigger2 = permBeginOfLineStringTrigger("dummy2", "", {"Does this work", "Does that work"}, "echo('Yes it does!\\n')")
  myTestFunctions.testTrigger3 = permBeginOfLineStringTrigger("dummy3", "MyTriggerGroup", {"Does this work"}, "echo('Yes it does!\\n')")
  myTestFunctions.testTrigger4 = permBeginOfLineStringTrigger("dummy4", "MyTriggerGroup", {"Does this work", "Does that work"}, "echo('Yes it does!\\n')")
end
```

(I was going to script a second function to clean the created things up but forgot that we don't allow permanent things to be deleted!)

Anyhow the behaviour in the two points in the Git History ***are the same***. The two triggers (`"dummy3"`  and `"dummy4"` in the `"MyTriggerGroup"` are created as "multi-item types" - i.e. without the `AND / Multi-line` groupbox checked - as is the single item root level trigger `"dummy1"` - the only one that DOES have the groupbox checked is `"dummy2"` with more than one item.

![image](https://github.com/user-attachments/assets/3ff989f2-9d19-4e77-b3a5-86927d2b8b2f)

Trying to reproduce the behaviour as described in https://github.com/Mudlet/Mudlet/issues/7480#issuecomment-2439628270 I got this from the development branch code `1fa61b1d8` - it produced a trigger with the `AND / Multi-line (delta)` checkbox enabled when done at the root level:
![image](https://github.com/user-attachments/assets/70db3225-b6c7-4a22-8c47-f58a2f70e389)
and inside a folder it does NOT:
![image](https://github.com/user-attachments/assets/7c91610a-105c-437a-baf6-b856323be6b2)

The code in this PR - when invoking the `perm*Trigger(...)` functions from the Mudlet command line also produced a Multi-line trigger when done at the root level:
![image](https://github.com/user-attachments/assets/bda923af-c8ee-4b3f-a3d0-e1d19b66ac91)
and a multi-item one when done in a trigger group:
![image](https://github.com/user-attachments/assets/f9237cdd-274b-4ee2-b59e-c06be38d1033)

Importantly I deleted the triggers from the first run after starting the second - so that the ID numbers should be different - so they can be verified from the output when run from the command line and from the editor views in the screenshots. As such I think the idea that this PR is not reproducing the prior behaviour when called without a fifth argument so as to reproduce past scripts that using the functions concerned is erroneous! It might be worth noting that the editor screen did not update immediately after running the `perm*Trigger` function from the command line - there was a noticeable delay of a fraction of a second.

vadi2 commented:
OK, lets go for it.
